internal/rbd/rbd_util.go

```diff
@@ -825,47 +811,30 @@ func getSnapshotMetadata(ctx context.Context, pSnapOpts *rbdSnapshot, cr *util.C
 	return nil
 }
 
-// imageInfo strongly typed JSON spec for image info
-type imageInfo struct {
-	ObjectUUID string   `json:"name"`
-	Size       int64    `json:"size"`
-	Features   []string `json:"features"`
-	CreatedAt  string   `json:"create_timestamp"`
-}
-
 // getImageInfo queries rbd about the given image and returns its metadata, and returns
 // ErrImageNotFound if provided image is not found
-func getImageInfo(ctx context.Context, monitors string, cr *util.Credentials, poolName, imageName string) (imageInfo, error) {
-	// rbd --format=json info [image-spec | snap-spec]
-
-	var imgInfo imageInfo
+func (rv *rbdVolume) getImageInfo() error {
+	image, err := rv.open()
+	if err != nil {
+		return err
+	}
+	defer image.Close()
 
-	stdout, stderr, err := util.ExecCommand(
-		"rbd",
-		"-m", monitors,
-		"--id", cr.ID,
-		"--keyfile="+cr.KeyFile,
-		"-c", util.CephConfigPath,
-		"--format="+"json",
-		"info", poolName+"/"+imageName)
+	imageInfo, err := image.Stat()
 	if err != nil {
-		klog.Errorf(util.Log(ctx, "failed getting information for image (%s): (%s)"), poolName+"/"+imageName, err)
-		if strings.Contains(string(stderr), "rbd: error opening image "+imageName+
-			": (2) No such file or directory") {
-			return imgInfo, ErrImageNotFound{imageName, err}
-		}
-		return imgInfo, err
+		return ErrImageNotFound{err: errors.Wrapf(err, "failed to get stat of image")}
 	}
+	// TODO: can rv.VolSize not be a uint64? Or initialize it to -1?
+	rv.VolSize = int64(imageInfo.Size)
 
-	err = json.Unmarshal(stdout, &imgInfo)
+	features, err := image.GetFeatures()
 	if err != nil {
-		klog.Errorf(util.Log(ctx, "failed to parse JSON output of image info (%s): (%s)"),
-			poolName+"/"+imageName, err)
-		return imgInfo, fmt.Errorf("unmarshal failed: %+v.  raw buffer response: %s",
-			err, string(stdout))
+		return ErrImageNotFound{err: errors.Wrapf(err, "failed to get features of image")}
```

Comment:
will this aways be imageNotFound Error?

Reply:
it is most likely some other error, as opening the image already worked. But the comment for this function mentions it should return ErrImageNotFound (although if `rv.open()` fails, it could be `util.ErrPoolNotFound`)

Reply:
with current implementation we are checking the image not found error from the rbd info command and returning the not found error if the image is not found, with the new implementation we need to return not found only if the image is not found. we need to do some check on the error return from the go-ceph and return a proper error message. returning ErrNotFound can leak the images in the backend.